### PR TITLE
Company details change request

### PR DESCRIPTION
**Pull request by rafenden**
## Description of change

This PR will update the company business details edit form for matched companies which after submission will create Zendesk tickets for each updated field.

## Test instructions

1. Go to a matched company
2. Go to the business details edit page
3. Change some field, e.g. "Trading names"
4. Check if a Zendesk ticket was created (check our staging zendesk instance)

### CODE REVIEW

File: src/apps/companies/apps/edit-company/client/OneListFields.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    sectors.find((s) => company.sector && company.sector.id === s.value),
+    'label'
+  )
+  const headquarterType = get(
+    headquarterTypes.find(
+      (s) => company.headquarter_type && company.headquarter_type.id === s.value
+    ),
+    'label',
+    undefined
+  )
```
Comment (web-bert) [inline src/apps/companies/apps/edit-company/client/OneListFields.jsx]: Same here, use optional chaining

File: src/apps/companies/apps/edit-company/controllers.js [hunk 2]
```diff
@@ -66,17 +61,77 @@ async function renderEditCompanyForm(req, res, next) {
...
+        changeRequests: zendeskResponses.map((r) => r.data.ticket.description),
+      })
+    } else {
+      const result = await updateCompany(
+        req.session.token,
+        company.id,
+        apiRequestFields
+      )
+      req.flash('success', 'Company record updated')
+      res.status(200).json(result)
```
Comment (web-bert) [inline src/apps/companies/apps/edit-company/controllers.js]: `.status(200)` is not needed here

File: src/apps/companies/apps/edit-company/transformers.js [hunk 3]
```diff
@@ -35,39 +56,44 @@ const transformCompanyToForm = (company) => {
...
-    isUndefined
+    (v, k) => isUndefined(v) || !whitelistedFields.includes(k)
+  )
+}
+
+const transformFormToZendesk = (company, formValues) => {
+  const originalFormValues = transformCompanyToForm(company)
+  return omitBy(
+    formValues,
+    (v, k) =>
```
Comment (web-bert) [inline src/apps/companies/apps/edit-company/transformers.js]: Can you use more descriptive names here for the variables?

File: test/sandbox/routes/zendesk.js [hunk 4]
```diff
@@ -1,5 +1,10 @@
 var zendeskTickets = require('../fixtures/zendesk/tickets.json')
 
 exports.tickets = function(req, res) {
-  res.json(201, JSON.stringify(zendeskTickets))
+  var response = _.set(
+    zendeskTickets,
+    'ticket.description',
+    req.body.ticket.comment.body
+  )
+  res.json(201, JSON.stringify(response))
```
Comment (web-bert) [inline test/sandbox/routes/zendesk.js]: This use of res.json is deprecated and the use of stringify is unnecessary, so this should be `res.status(201).json(response)`

File: test/functional/cypress/support/assertions.js [hunk 5]
```diff
@@ -108,28 +102,14 @@ const assertFieldSelect = ({
     .should('have.length', optionsCount)
     .then(
       () =>
-        value &&
+        value ??
```
Comment (ian-leggett) [inline test/functional/cypress/support/assertions.js]: Nice

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx [hunk 6]
```diff
@@ -80,23 +79,15 @@ const CompanyBusinessDetails = ({
...
-            . This is an external and verified source. The information is
-            automatically updated.{' '}
-            <Link href={urls.companyEditHistory}>View the edit history</Link>.
-          </p>
-          <p>
-            If you think the information is incomplete or incorrect,{' '}
-            <Link href={urls.support}>get in touch using the support form</Link>
-            .
-          </p>
+          Most business details have been verified by trusted third-parties to
```
Comment (ian-leggett) [inline src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: Should this text be wrapped in `<Paragraph>`?

File: src/apps/companies/apps/business-details/client/SectionAbout.jsx [hunk 7]
```diff
@@ -32,9 +32,7 @@ const SectionAbout = ({ businessDetails, isDnbCompany, isArchived, urls }) => (
   <SummaryTable
     caption={`About ${businessDetails.name}`}
     data-auto-id="aboutDetailsContainer"
-    actions={
-      !isDnbCompany && !isArchived && <Link href={urls.companyEdit}>Edit</Link>
-    }
+    actions={!isArchived && <Link href={urls.companyEdit}>Edit</Link>}
```
Comment (ian-leggett) [inline src/apps/companies/apps/business-details/client/SectionAbout.jsx]: Could we use `??` ?

File: src/apps/companies/apps/edit-company/client/CommonFields.jsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      type="text"
+    />
+
+    <FieldAddress
+      legend="Address"
+      name="address"
+      hint="This should be the address for this particular office of the
+         business. If you need to record activity or a contact for a different
+          address, please add a new company record to Data Hub."
+      country={get(company, 'address.country', {})}
```
Comment (ian-leggett) [inline src/apps/companies/apps/edit-company/client/CommonFields.jsx]: If you remove lodash here you could use ```company?.address?.country ?? {}```

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx [hunk 9]
```diff
@@ -80,23 +79,15 @@ const CompanyBusinessDetails = ({
 
       {isDnbCompany && (
         <Details
-          summary="Where does information on this page come from?"
+          summary="Are these business details right?"
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: This was suggested by the content designer.

File: src/apps/companies/apps/edit-company/client/CommonFields.jsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    <FieldInput
+      label="Business description (optional)"
+      name="description"
+      type="text"
+    />
+
+    <FieldAddress
+      legend="Address"
+      name="address"
+      hint="This should be the address for this particular office of the
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/CommonFields.jsx]: This content was supplied by the content designer.

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx (hunk 6 above)
Comment (rafenden) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: I didn't see a need for additional HTML there.

File: src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx [hunk 11]
```diff
@@ -261,7 +78,8 @@ function EditCompanyForm({
 EditCompanyForm.propTypes = {
   isOnOneList: PropTypes.bool.isRequired,
   csrfToken: PropTypes.string.isRequired,
-  companyDetails: PropTypes.object.isRequired,
+  company: PropTypes.shape({}).isRequired,
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/EditCompanyForm.jsx]: Good point, the code was updated.

File: src/apps/companies/apps/edit-company/client/CompanyMatched.jsx [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+      />
+    )}
+
+    <H4 as="h2">After you click submit</H4>
+
+    <Paragraph>
+      Changes will be reviewed by our third-party data supplier and updated.
+      Business description, region and sector are not updated by third parties.
+    </Paragraph>
+  </>
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/CompanyMatched.jsx]: That was the design, which is also consistent with the company matching form.

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx (hunk 6 above)
Comment (rafenden) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: The `Paragraph` component is a tricky one, it only accepts markdown content as children.

Comment (rafenden) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: Used `p` instead, thanks :)

File: src/apps/companies/apps/edit-company/client/OneListFields.jsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+import { get } from 'lodash'
+import Details from '@govuk-react/details'
+import Link from '@govuk-react/link'
+import { FieldUneditable } from 'data-hub-components'
+
+function OneListFields({ company, sectors, headquarterTypes, oneListEmail }) {
+  const sectorName = get(
+    sectors.find((s) => company.sector && company.sector.id === s.value),
+    'label'
+  )
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/OneListFields.jsx]: Good idea, it's much nicer now, thanks!

File: src/apps/companies/apps/edit-company/client/OneListFields.jsx (hunk 1 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/OneListFields.jsx]: Done, thanks!

File: src/apps/companies/apps/edit-company/transformers.js (hunk 3 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/transformers.js]: Good point, var names updated!

File: test/sandbox/routes/zendesk.js (hunk 4 above)
Comment (rafenden) [reply test/sandbox/routes/zendesk.js]: Good point, the code was updated.

File: src/apps/companies/apps/edit-company/controllers.js (hunk 2 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/controllers.js]: Thanks, `status(200)` removed.

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx (hunk 6 above)
Comment (ian-leggett) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: Ah yeah, the same thing has happened to me this morning with `<Paragraph>`

File: src/apps/companies/apps/edit-company/client/CommonFields.jsx [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+import { get } from 'lodash'
+import { WEBSITE_REGEX } from '../../add-company/client/constants'
+
+const websiteValidator = (value) =>
+  value && !WEBSITE_REGEX.test(value) ? 'Enter a valid website URL' : null
+
+const CommonFields = ({ company, regions }) => (
+  <>
+    <FieldInput
+      label="Trading name (optional)"
```
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/CommonFields.jsx]: Thanks, I updated the label.

File: src/apps/companies/apps/edit-company/client/CommonFields.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/CommonFields.jsx]: This is the content from the existing company edit form which didn't raise any concerns with the content designer, so I kept is as-is.

File: src/apps/companies/apps/edit-company/client/CommonFields.jsx (hunk 8 above)
Comment (rafenden) [reply src/apps/companies/apps/edit-company/client/CommonFields.jsx]: Code updated, the new chain operator is really nice!

File: src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx (hunk 6 above)
Comment (rafenden) [reply src/apps/companies/apps/business-details/client/CompanyBusinessDetails.jsx]: Good point, used `p`, thanks!

**Pull request merged**